### DISCUSSION TITLE
Replicating @babel/template functionality

### DISCUSSION THREAD

**Discussion by lewisl9029** [swc-project/swc#3464] [Q&A] [answered]
Hi there!

I have a custom babel plugin that currently uses the [`@babel/template`](https://babel.dev/docs/en/babel-template) package to inject blocks of code into files without breaking sourcemaps. A vastly simplified version looks something like this:

```js
const createPlugin = () => {
  // The console.logs are just placeholders, the actual code is quite long
  // I'd like to avoid manually rewriting into ASTs if possible
  const beforeAst = babelTemplate.ast(`console.log('hi!')`)

  const afterAst = babelTemplate.ast(`console.log('bye!')`)

  return () => ({
    visitor: {
      Program: {
        exit: (path) => {
          path.unshiftContainer('body', beforeAst)
          path.pushContainer('body', afterAst)
        },
      },
    },
  })
}
```

I'd love to migrate it over to swc for those sweet performance gains, but after looking through the docs, I'm still unsure how I could accomplish the same task. 

Any assistance would be very much appreciated!

**Comment by kdy1** [comment, marked as answer]
Alternative for `@babel/template` is WIP.

**Reply by lewisl9029** [reply]
That's amazing! Thank you so much! 

Just found the PR here for anyone else who might be looking for it: https://github.com/swc-project/swc/pull/3155

Will follow along!

**Reply by bryanjtc** [reply]
What is the status of the alternative? I see that #3155 was merged, but no mention of how to replace @babel/template with swc in any of the docs.

**Reply by kdy1** [reply]
See https://rustdoc.swc.rs/swc_ecma_quote/macro.quote.html for usage

**Comment by redexp** [comment]
@kdy1 `quote!` is looks like replacement for babel template, but it can handle only one expression. It could be full replacement if `quote` handle full `Module` not just `ModuleItem`.
